src/com/Communication.cpp

```diff
@@ -38,28 +39,30 @@ void Communication::connectMasterSlaves(std::string const &participantName,
 /**
  * @attention This method modifies the input buffer.
  */
-void Communication::reduceSum(double const *itemsToSend, double *itemsToReceive, int size)
+void Communication::reduceSum(precice::span<double const> itemsToSend, precice::span<double> itemsToReceive)
 {
-  PRECICE_TRACE(size);
+  PRECICE_TRACE(itemsToSend.size(), itemsToReceive.size());
+  PRECICE_ASSERT(itemsToSend.size() == itemsToReceive.size());
 
-  std::copy(itemsToSend, itemsToSend + size, itemsToReceive);
+  std::copy(itemsToSend.begin(), itemsToSend.end(), itemsToReceive.begin());
 
-  std::vector<double> received(size);
+  std::vector<double> received(itemsToReceive.size());
   // receive local results from slaves
   for (size_t rank = 0; rank < getRemoteCommunicatorSize(); ++rank) {
```

Comment:
IIRC I had already a discussion with @fsimonis and we agreed on using `++rank` instead of `++rank`

Reply:
@DavidSCN you wrote `++rank` twice. On what did you agree? As long as it's consistent I'm fine. Personally `rank++` is more natural to be, but I don't really mind.

Reply:
Oops, corrected it :D
Yes, more naturally, but has some disadvantage I don't remember.

Reply:
`++thing` increments the thing and returns a reference to the thing.

`thing++` saves a copy, then increments the thing and then returns the copy. Using this without using the return value doesn't really make sense.

Compilers know how to get rid of the unnecessary work in most cases. Hence this doesn't matter for most cases. Most cases is not every case though. Imagine some fancy iterator that does some extra work on advance like evaluating a function or filtering something.

So, if `++thing` is what you want to do, you should use it. Using even in cases it doesn't matter simply is a matter of consistency.

Reply:
Sounds like a good reason for keeping `++thing`. I will try to get used to `++thing`, as well, and change my habits.

Interesting: In the CPP reference a for loop also uses `++thing` (https://en.cppreference.com/w/cpp/language/for). So I have no idea, why the `thing++` is more natural for me in this context.

Reply:
> So I have no idea, why the `thing++` is more natural for me in this context.

It is more natural because we see it more often (see... C++).

To demystify this, look at the signatures of the methods in the table at the top of this page: https://en.cppreference.com/w/cpp/language/operator_incdec The `thing++` takes an argument by value, returns a value. The `++thing` takes a reference and returns a reference. As @fsimonis wrote, in most cases it does not matter, but there can be specific cases where it does.

There seems to be no [core guideline](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md) for this. I would just try to use `++thing` by default for the same reason as we are using references by default: safety (no accidental slicing of objects) and efficiency (no unnecessary copying around).